### PR TITLE
feat: better error message for invalid induction alternative name

### LINKED ISSUE #5887: RFC: Improved error msg for induction alternative names
### Proposal

If the alternative names for the induction tactic are incorrect,  it just returns: "invalid alternative name 'name'".
Here is one example:
image
It would be very nice, if the correct alternative names could be proposed direct.

Note, that in the example above the induction is done over  \mathbb{Z} instead of \mathbb{N} and hence other alternative names are required.

  - **User Experience**: How does this feature improve the user experience?
      - Improves the error message and allows users to quickly fix their bugs

  - **Beneficiaries**: Which Lean users and projects benefit most from this feature/change?
     - Mathlib users will get additional information that is useful for proof writing
     - AI bots will get this message and can quicker iterate

  - **Maintainability**: Will this change streamline code maintenance or simplify its structure?
     - No additional burden on maintainablity

### Community Feedback

No feedback available.

### Impact

Add :+1: to [issues you consider important](https://github.com/leanprover/lean4/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc). If others benefit from the changes in this proposal being added, please ask them to add :+1: to it.

### PR DESCRIPTION
**Pull request by josojo**
Closes #5887

### CODE REVIEW

File: src/Lean/Elab/Tactic/Induction.lean [hunk 1]
```diff
@@ -216,7 +216,11 @@ private def checkAltNames (alts : Array Alt) (altsSyntax : Array Syntax) : Tacti
     let altName := getAltName altStx
     if altName != `_ then
       unless alts.any (·.name == altName) do
-        throwErrorAt altStx "invalid alternative name '{altName}'"
+        let validNames := (alts.map (·.name)).toList |>List.take 3
+        let validNamesStr := validNames.foldl (init := "") fun acc n =>
+          if acc == "" then s!"{Name.toString n}" else s!"{acc}, {Name.toString n}"
+        throwErrorAt altStx s!"invalid alternative name '{altName}', valid options include: {validNamesStr}"
```
Comment (nomeata) [inline src/Lean/Elab/Tactic/Induction.lean:222]: Thanks for looking into this. I suggest to do a bit more while you are at it, and make the `checkAltNames` function check

* That `_` is last, if it exists.
* That no name is duplicate.
* When an name is not found in `alts`, say “invalid alternative name '{altName}', expected one of …, … or, …“, and list all those alternative names that are not explicitly handled already. (If this list is empty, print a different message, e.g. “…, no unhandled alternatives”.

Comment (josojo) [reply src/Lean/Elab/Tactic/Induction.lean:222]: >That _ is last, if it exists.

this is done already in the code, right?
```
if getAltName altStx == `_ && i != altsSyntax.size - 1 then
      withRef altStx <| throwError "invalid occurrence of wildcard alternative, it must be the last alternative"
```
happy to tackle the other two suggestions.

File: src/Lean/Elab/Tactic/Induction.lean [hunk 2]
```diff
@@ -208,18 +208,33 @@ private def getAltNumFields (elimInfo : ElimInfo) (altName : Name) : TermElabM N
...
+        else
+          let names := unhandledAlts.map (·.name)
+          let namesStr := match names.size with
+            | 0 => ""
+            | 1 => s!"{Name.toString names[0]!}"
+            | 2 => s!"{Name.toString names[0]!} or {Name.toString names[1]!}"
+            | _ =>
+              let init := names.extract 0 (names.size - 1)
+              let last := names.back
+              s!"{", ".intercalate (init.toList.map Name.toString)}, or {last}"
```
Comment (nomeata) [inline src/Lean/Elab/Tactic/Induction.lean:235]: This code obscures the intent here. Maybe define a variant of [Lean.MessageData.andList](https://leanprover-community.github.io/mathlib4_docs/Lean/Message.html#Lean.MessageData.andList) (in that that module) and use it here.

Also omit the `one of` below, so that the grammar works out with even one left.

File: tests/lean/inductionErrors.lean [hunk 3]
```diff
@@ -79,3 +79,9 @@ theorem ex11 (p q : Nat) : p ≤ q ∨ p > q := by
   | upper d => apply Or.inr; admit
   | lower d /- error duplicate -/ => apply Or.inl; admit
   | diag    => apply Or.inl; apply Nat.le_refl
+
+theorem ex12 (p: Nat) : p ≤ q ∨ p > q := by
+  cases p, q using elimEx with
+  | lower2 /- error -/ d => apply Or.inl; admit
+  | upper d => apply Or.inr
+  | diag    => apply Or.inl; apply Nat.le_refl
```
Comment (nomeata) [inline tests/lean/inductionErrors.lean:87]: These tests with the `.expected.out` output are a bit out of fashion. Can you add a new test file in the `tests/lean/run` directory and use `#guard_msgs in` to assert the expected output? Also makes it easier to review.

File: tests/lean/inductionErrors.lean.expected.out [hunk 4]
```diff
@@ -25,7 +25,8 @@ y : Nat
 ⊢ 0 + (y + 1) = y + 1
 inductionErrors.lean:50:2-50:16: error: alternative 'cons' is not needed
 inductionErrors.lean:55:2-55:16: error: alternative 'cons' is not needed
-inductionErrors.lean:60:2-60:40: error: invalid alternative name 'upper2'
+inductionErrors.lean:60:2-60:40: error: invalid alternative name 'upper2', expected one of diag or upper
 inductionErrors.lean:66:2-66:28: error: invalid occurrence of wildcard alternative, it must be the last alternative
 inductionErrors.lean:74:2-74:34: error: unused alternative '_'
-inductionErrors.lean:80:2-80:56: error: unused alternative 'lower'
+inductionErrors.lean:80:2-80:56: error: duplicate alternative name 'lower'
+inductionErrors.lean:85:2-85:47: error: invalid alternative name 'lower2', expected one of diag, upper, or lower
```
Comment (david-christiansen) [inline tests/lean/inductionErrors.lean.expected.out:32]: The alternative names should be in single quotes here, like:
```suggestion
inductionErrors.lean:85:2-85:47: error: invalid alternative name 'lower2', expected one of 'diag', 'upper', or 'lower'
```

File: src/Lean/Elab/Tactic/Induction.lean [hunk 5]
```diff
@@ -208,15 +208,28 @@ private def getAltNumFields (elimInfo : ElimInfo) (altName : Name) : TermElabM N
...
-        throwErrorAt altStx "invalid alternative name '{altName}'"
+        let unhandledAlts := alts.filter fun alt => !seenNames.contains alt.name
+        let msg ← do
+          if unhandledAlts.isEmpty then
+            pure s!"invalid alternative name '{altName}', no unhandled alternatives"
+          else
+            let unhandledAltsMessages := ((unhandledAlts.map (·.name)).map MessageData.ofName).toList
+            let unhandledAltsStr ← MessageData.toString (MessageData.orList unhandledAltsMessages)
+            pure s!"invalid alternative name '{altName}', expected {unhandledAltsStr}"
+        throwErrorAt altStx msg
```
Comment (nomeata) [inline src/Lean/Elab/Tactic/Induction.lean:231]: Probably doesn’t matter much, but no need to go through `String` here, is there? Then this becomes non-monadic. Also, repeated `map` can be a single map.

Untested:

```suggestion
        let msg :=
          if unhandledAlts.isEmpty then
            m!"invalid alternative name '{altName}', no unhandled alternatives"
          else
            let unhandledAltsMessages := unhandledAlts.map (m!"{·.name}")
            let unhandledAlts := MessageData.orList unhandledAltsMessages.toList
            m!"invalid alternative name '{altName}', expected {unhandledAlts}"
        throwErrorAt altStx msg
```

File: tests/lean/run/inductionCheckAltNames.lean [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+  (lower : (delta a : Nat) → motive (a + delta.succ) a)
+  : motive y x
+
+/-- error: invalid alternative name 'lower2', expected 'diag', 'upper' or 'lower' -/
+#guard_msgs in
+theorem invalidAlt (p: Nat) : p ≤ q ∨ p > q := by
+  cases p, q using elimEx with
+  | lower2 /- error -/ d => apply Or.inl; admit
+  | upper d => apply Or.inr
+  | diag    => apply Or.inl; apply Nat.le_refl
```
Comment (nomeata) [inline tests/lean/run/inductionCheckAltNames.lean:15]: Good! Can you add test cases for all other interesting variants (duplicates, unexpected with none left, unexpected with one left, maybe duplicate + unexpected)?

**Review by nomeata — COMMENTED** [review]
Almost there :-)
